### PR TITLE
Panning and zooming

### LINKED ISSUE #27: Zoom

### LINKED ISSUE #28: Panning

### PR DESCRIPTION
**Pull request by Alexander-Ploskin**
UI:  [imgur](https://imgur.com/a/eKVT9Hj)

Completed:

- [x] panning
- [x] fix landmark hover when mouse leaves frame
- [x] zooming

### CODE REVIEW

File: src/main/kotlin/sliv/tool/scene/view/KeypointView.kt [hunk 1]
```diff
@@ -14,11 +13,11 @@ class KeypointView(private val keypoint: Landmark.Keypoint) : LandmarkView() {
 
     private var radius = OrdinaryRadius
 
-    override fun draw(gc: GraphicsContext) {
+    override fun draw(gc: GraphicsContext, scale: Double) {
         gc.fill = keypoint.layer.color
         gc.globalAlpha = keypoint.layer.opacity
-        val x = keypoint.coordinate.x.toDouble() - radius / 2
-        val y = keypoint.coordinate.y.toDouble() - radius / 2
+        val x = keypoint.coordinate.x.toDouble() * scale - radius / 2
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/KeypointView.kt:21]: what about radius multiplication?

File: src/main/kotlin/sliv/tool/scene/view/KeypointView.kt [hunk 2]
```diff
@@ -28,14 +27,9 @@ class KeypointView(private val keypoint: Landmark.Keypoint) : LandmarkView() {
...
-        val xDiff = abs(keypoint.coordinate.x - event.x)
-        val yDiff = abs(keypoint.coordinate.y - event.y)
-
-        state = if (hypot(xDiff, yDiff) < radius) {
-            LandmarkState.Hovered
-        } else {
-            LandmarkState.Ordinary
-        }
+    override fun isHovered(x: Double, y: Double, scale: Double): Boolean {
+        val xDiff = abs(keypoint.coordinate.x - x / scale)
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/KeypointView.kt:33]: keypoint positions are already updated with scale?

Comment (DmiitriyJarosh) [reply src/main/kotlin/sliv/tool/scene/view/KeypointView.kt:33]: why is it not multiplied?

File: src/main/kotlin/sliv/tool/scene/view/PlaneView.kt [hunk 3]
```diff
@@ -1,26 +1,25 @@
...
 class PlaneView(private val plane: Landmark.Plane) : LandmarkView() {
-    override fun draw(gc: GraphicsContext) {
+    override fun draw(gc: GraphicsContext, scale: Double) {
         gc.fill = Color.RED //TODO: use some colorizing delegate here
         gc.globalAlpha = plane.layer.opacity
         when (state) {
             LandmarkState.Ordinary -> {
                 plane.points.forEach {
-                    gc.fillRect(it.x.toDouble(), it.y.toDouble(), 1.0, 1.0)
+                    gc.fillRect(it.x.toDouble() * scale, it.y.toDouble() * scale, 1.0 * scale, 1.0 * scale)
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/PlaneView.kt]: why `1.0` is required?

File: src/main/kotlin/sliv/tool/scene/view/SceneView.kt [hunk 4]
```diff
@@ -28,22 +31,51 @@ class SceneView : View() {
...
+            if(event.isConsumed) { // If event is consumed by vsp
+                return@setOnScroll
+            }
+
+            val initialPos = grid.getPosition()
+
+            val initialMouseX = (initialPos.first + event.x) / scaleProperty.value
+            val initialMouseY = (initialPos.second + event.y) / scaleProperty.value
+
+            if(event.deltaY > 0) {
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/SceneView.kt]: why do you skip spaces after `if`?

File: src/main/kotlin/sliv/tool/scene/view/SceneView.kt [hunk 5]
```diff
@@ -28,22 +31,51 @@ class SceneView : View() {
...
+                scaleProperty.value = max(scaleProperty.value / scaleFactor, minScale)
+            }
+
+            val translatedMouseX = initialMouseX * scaleProperty.value
+            val translatedMouseY = initialMouseY * scaleProperty.value
+
+            grid.scrollTo(translatedMouseX - event.x, translatedMouseY - event.y)
+        }
+
+        add(grid.getNode())
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/SceneView.kt]: to property syntax?

File: src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGrid.kt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    private var initialValues = Position.of(0.0, 0.0)
+
+    override fun setUpPanning() {
+        virtualGrid.setOnMousePressed { event ->
+            dragStart = Position.of(event.x, event.y)
+            initialValues = virtualGrid.position
+        }
+
+        virtualGrid.setOnMouseDragged { event ->
+            val xDelta = -(event.x - dragStart.x)
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGrid.kt]: maybe just change arguments and remove outer `-`?

File: src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGrid.kt [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+            val yDelta = -(event.y - dragStart.y)
+            virtualGrid.scrollTo(initialValues.x + xDelta, Orientation.HORIZONTAL)
+            virtualGrid.scrollTo(initialValues.y + yDelta, Orientation.VERTICAL)
+        }
+    }
+
+    override fun setOnScroll(handler: EventHandler<ScrollEvent>) {
+        virtualGrid.addEventHandler(ScrollEvent.SCROLL, handler)
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGrid.kt:39]: what about side scrollers that can be used without mouse wheel? are they hidden?

File: src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGrid.kt [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    override fun scrollTo(x: Double, y: Double) {
+        virtualGrid.scrollTo(x, Orientation.HORIZONTAL)
+        virtualGrid.scrollTo(y, Orientation.VERTICAL)
+    }
+
+    override fun getPosition(): Pair<Double, Double> {
+        return Pair(virtualGrid.position.x, virtualGrid.position.y)
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGrid.kt]: use `to` operator

File: src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGridProvider.kt [hunk 9]
```diff
@@ -5,26 +5,45 @@ import io.github.palexdev.mfxcore.collections.ObservableGrid
...
+    ): Grid {
         val gridData = ObservableGrid.fromList(data, columnsNumber)
         val grid = VirtualGrid(gridData) { item -> FrameViewAdapter(cellFactory(item)) }
         grid.cellSize = Size(cellWidth, cellHeight)
 
         val vsp = VSPUtils.wrap(grid)
         vsp.layoutMode = ScrollPaneEnums.LayoutMode.COMPACT
         vsp.isAutoHideBars = true
+        vsp.isSmoothScroll = true
+        VSPUtils.setVSpeed(vsp, 100.0, 100.0, 100.0)
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGridProvider.kt:32]: why is this required? maybe some comment?

File: src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGridProvider.kt [hunk 10]
```diff
@@ -5,26 +5,45 @@ import io.github.palexdev.mfxcore.collections.ObservableGrid
...
 
-        return vsp
+        scale.onChange { newScale ->
+            // Virtual grid breaks if cell size changes when x position differs from 0.0
+            // Virtual grid always drops position to (0, 0) when cell size changes, so this solution is ok
+            grid.scrollTo(0.0, Orientation.HORIZONTAL)
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGridProvider.kt:37]: is scroll reseted here?

Comment (DmiitriyJarosh) [reply src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGridProvider.kt:37]: how it will be restored?

**Review by DmiitriyJarosh — CHANGES_REQUESTED** [review]

File: src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGridProvider.kt (hunk 9 above)
Comment (Alexander-Ploskin) [reply src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGridProvider.kt:32]: Vertical scrolling is too fast by default. It's needed to be tuned to achieve smooth scrolling

File: src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGrid.kt (hunk 7 above)
Comment (Alexander-Ploskin) [reply src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGrid.kt:39]: I consume scroll event by VirtualGrid to prevent scrolling when mouse is on frames. When mouse is on scrollbars they appears and handles scroll by itself, so mouse wheel provokes scrolling.

File: src/main/kotlin/sliv/tool/scene/view/KeypointView.kt (hunk 2 above)
Comment (Alexander-Ploskin) [reply src/main/kotlin/sliv/tool/scene/view/KeypointView.kt:33]: I calculate distance in absolute coordinate system independent from scale, so I divide x by scale.

File: src/main/kotlin/sliv/tool/scene/view/KeypointView.kt (hunk 1 above)
Comment (Alexander-Ploskin) [reply src/main/kotlin/sliv/tool/scene/view/KeypointView.kt:21]: I thought that features shouldn't change its size to look better on big scales (it's very disappointing to see keypoint mark 80 x 80 pixels)

File: src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGridProvider.kt (hunk 10 above)
Comment (Alexander-Ploskin) [reply src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGridProvider.kt:37]: I reset scroll to prevent exception from VirtualGrid. It doesn't change anything, because on cell size changing VirtualScrollPane resets scroll. Position is restored in zoomGrid method (SceneView.kt).

File: src/main/kotlin/sliv/tool/scene/view/KeypointView.kt (hunk 1 above)
Comment (DmiitriyJarosh) [reply src/main/kotlin/sliv/tool/scene/view/KeypointView.kt:21]: yes, but when you zoom out, it looks like one red mark on the whole image. Maybe some ideas how to make it better?

File: src/main/kotlin/sliv/tool/scene/view/Grid.kt [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+interface Grid {
+    fun setUpPanning()
+
+    val node: Node
+
+    fun setOnScroll(handler: EventHandler<ScrollEvent>)
+
+    fun scrollTo(x: Double, y: Double)
+
+    fun getPosition(): Pair<Double, Double>
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/Grid.kt]: maybe currentPosition and property?

File: src/main/kotlin/sliv/tool/scene/view/SceneView.kt [hunk 12]
```diff
@@ -28,22 +35,59 @@ class SceneView : View() {
...
         }
 
-        add(grid)
+        grid.setUpPanning()
+
+        grid.setOnScroll { event ->
+            if (event.isConsumed) { // If event is consumed by vsp
+                return@setOnScroll
+            }
+            if(event.deltaY.compareTo(0) == 0) {
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/SceneView.kt]: compareTo? why simple == is not enough?

File: src/main/kotlin/sliv/tool/scene/view/SceneView.kt [hunk 13]
```diff
@@ -28,22 +35,59 @@ class SceneView : View() {
...
+                return@setOnScroll
+            }
+            zoomGrid(scaleProperty, grid, event.x to event.y, event.deltaY > 0)
+        }
+
+        add(grid.node)
+    }
+
+    private fun zoomGrid(scaleProperty: DoubleProperty, grid: Grid, mousePos: Pair<Double, Double>, isPositive: Boolean) {
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/SceneView.kt]: dont short words `mousePosition`

File: src/main/kotlin/sliv/tool/scene/view/SceneView.kt [hunk 14]
```diff
@@ -28,22 +35,59 @@ class SceneView : View() {
...
+        val initialPos = grid.getPosition()
+
+        val initialMouseX = (initialPos.first + mousePos.first) / scaleProperty.value
+        val initialMouseY = (initialPos.second + mousePos.second) / scaleProperty.value
+
+        if (isPositive) {
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/SceneView.kt:73]: why we need this? can we just check both borders instead?

File: src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGrid.kt [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+import javafx.scene.Node
+import javafx.scene.input.ScrollEvent
+import sliv.tool.scene.model.VisualizationFrame
+import sliv.tool.scene.view.Grid
+
+class VirtualizedFXGrid(
+    private val virtualGrid: VirtualGrid<VisualizationFrame, FrameViewAdapter>, private val vsp: VirtualScrollPane
+) : Grid {
+    private var dragStart = Position.of(-1.0, -1.0)
+    private var initialValues = Position.of(0.0, 0.0)
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGrid.kt]: what is it? maybe better name?

Comment (DmiitriyJarosh) [reply src/main/kotlin/sliv/tool/scene/view/virtualizedfx/VirtualizedFXGrid.kt]: dragStartGridPosition?

**Review by DmiitriyJarosh — CHANGES_REQUESTED** [review]

File: src/main/kotlin/sliv/tool/scene/view/KeypointView.kt (hunk 1 above)
Comment (Alexander-Ploskin) [reply src/main/kotlin/sliv/tool/scene/view/KeypointView.kt:21]: I make radius scaling, but only if scale < 1 to fix "one red mark on the whole image" problem.

File: src/main/kotlin/sliv/tool/scene/view/SceneView.kt (hunk 12 above)
Comment (Alexander-Ploskin) [reply src/main/kotlin/sliv/tool/scene/view/SceneView.kt]: It doesn't work with Double values

File: src/main/kotlin/sliv/tool/scene/view/SceneView.kt (hunk 14 above)
Comment (Alexander-Ploskin) [reply src/main/kotlin/sliv/tool/scene/view/SceneView.kt:73]: The body of this if statement not only checks the border, it also changes value for the scale property, so this method should know if scale should grow or decrease

File: src/main/kotlin/sliv/tool/scene/view/KeypointView.kt [hunk 16]
```diff
@@ -14,11 +13,13 @@ class KeypointView(private val keypoint: Landmark.Keypoint) : LandmarkView() {
...
     private var radius = OrdinaryRadius
 
-    override fun draw(gc: GraphicsContext) {
+    override fun draw(gc: GraphicsContext, scale: Double) {
         gc.fill = keypoint.layer.color
         gc.globalAlpha = keypoint.layer.opacity
-        val x = keypoint.coordinate.x.toDouble() - radius / 2
-        val y = keypoint.coordinate.y.toDouble() - radius / 2
+
+        radius = if (scale < 1) (OrdinaryRadius * scale) else OrdinaryRadius
```
Comment (DmiitriyJarosh) [inline src/main/kotlin/sliv/tool/scene/view/KeypointView.kt]: why are brackets required in ` (OrdinaryRadius * scale)`?

**Review by DmiitriyJarosh — APPROVED** [review]
